gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/layouting/CentroidFRLayouter.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright © 2014 - 2019 Leipzig University (Database Research Group)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradoop.flink.model.impl.operators.layouting;
+
+import org.apache.flink.api.java.DataSet;
+import org.apache.flink.api.java.operators.IterativeDataSet;
+import org.gradoop.common.model.impl.pojo.EPGMEdge;
+import org.gradoop.common.model.impl.pojo.EPGMVertex;
+import org.gradoop.flink.model.impl.epgm.LogicalGraph;
+import org.gradoop.flink.model.impl.functions.epgm.Id;
+import org.gradoop.flink.model.impl.operators.layouting.functions.AverageVertexPositionsFunction;
+import org.gradoop.flink.model.impl.operators.layouting.util.Centroid;
+import org.gradoop.flink.model.impl.operators.layouting.functions.CentroidRepulsionForceMapper;
+import org.gradoop.flink.model.impl.operators.layouting.functions.CentroidUpdater;
+import org.gradoop.flink.model.impl.operators.layouting.functions.FRRepulsionFunction;
+import org.gradoop.flink.model.impl.operators.layouting.functions.LVertexEPGMVertexJoinFunction;
+import org.gradoop.flink.model.impl.operators.layouting.util.Force;
+import org.gradoop.flink.model.impl.operators.layouting.util.GraphElement;
+import org.gradoop.flink.model.impl.operators.layouting.util.LEdge;
+import org.gradoop.flink.model.impl.operators.layouting.util.LGraph;
+import org.gradoop.flink.model.impl.operators.layouting.util.LVertex;
+import org.gradoop.flink.model.impl.operators.layouting.util.Vector;
+
+/**
+ * Layout a graph using approximate repulsive forces calculated using centroids as described
+ * <a href="https://www.researchgate.net/publication/281348264_Distributed_Graph_Layout_with_Spark">
+ * here</a>
+ * Very fast, even for large inputs.
+ */
+public class CentroidFRLayouter extends FRLayouter {
```

Comment:
I just created a couple of graph layouts for the small quickstart graph in our wiki. Although a layout is created, i observed some strange effects.
  1) The layout for an iteration count > 10 is almost always just a squashed mess of vertices in the upper left corner
  2) there seems to be a general tendency of the algorithm to just keep pushing the vertices in the upper left corner
For a small amount of iterations (< 5) the layout was readable but then it was not very aesthetic and resembled a random embedding? Maybe there is something wrong with the centroid/center update methods? or maybe its just the size/structure of the graph which does not fit this procedure. sadly i dont have any datasets at hand which are greater than the quickstart example but also smaller than the 3M vertex ldbc1.

Reply:
You are right. The CentroidFRLayouter does not seem to work with very small DataSets like the quickstart example. I also checked an older version, to make sure this is not a bug introduced by the refractoring (it is not).
Given how the CentroidLayouter works I am not too surprised it has trouble with such small graphs.
And yes, for a really small number of iterations the final layout is still mostly random.

The SuiteSparse Matrix Collection (https://sparse.tamu.edu/) is a great source for datasets and the MtxDataSource enables you to use them with gradoop (if you download the file-format called "matrix market").

For example the Dataset https://sparse.tamu.edu/MM/Hamm/add32.tar.gz works neatly with the CentroidFRLayouter

Reply:
Thanks thats very helpful! Given the fact that the centroids are a way to compute repulsive forces on distributed system in heuristic fashion, one could argue that the algorithm really does not need to be applied to such small graphs which can easily be processed by a single machine.

I'm definitely going to check the matrix collection out :+1: